### DISCUSSION TITLE
Compress dataportal requests/response

### DISCUSSION THREAD

**Discussion by BlagoCuljak** [MarimerLLC/csla#1694] [Questions] [answered]
Hi, I was wondering did anyone tried to compress 5.2 DP requests? I query the portal from WPF app, and for past couple of hours tried everything from [MS Docs](https://docs.microsoft.com/en-us/aspnet/core/performance/response-compression?view=aspnetcore-3.1) and still no compression on end of it.

Dynamic and static compression are activated on IIS 10.

I use CSLA 5.2 and httpdataportal.

**Comment by rockfordlhotka** [comment]
This isn't something I've explored.

The `HttpProxy` uses `HttpClient`, and maybe there is some client-side config you need to do on the `HttpClient` object to enable compression?

**Comment by BlagoCuljak** [comment]
![CSLA fetch](https://user-images.githubusercontent.com/34435929/84860524-7809fb80-b06f-11ea-85b0-5a81dec3b28d.png)

Hi Rocky,
The main issue is that i don't get any Content-Type, and therefore, I can't get the server rules to enable compression if eg. text/xml as it was with WCF.

I think that we need to add some thing here so that server and client can compress/decompress.
[https://tools.ietf.org/html/rfc7231#section-3.1.1.5](https://tools.ietf.org/html/rfc7231#section-3.1.1.5)

**Reply by rockfordlhotka** [reply]
That's a header value right? So we probably need to change [where the `httpRequest` is created](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.Shared/DataPortalClient/HttpProxy.cs#L417) to add it to the request header?

```c#
      httpRequest.Headers.Add("context-type", "application/binary");
```

(that's probably wrong - just brainstorming)

**Reply by rockfordlhotka** [reply]
This makes me think that we probably need an extensibility point here as well (in addition to setting the context-type always). I can imagine scenarios where people would need to add other custom header elements for SSO or other things? 

No one has asked for such a thing though, so maybe I'm anticipating scenarios that don't exist.

**Reply by BlagoCuljak** [reply]
Yes, as you can see, I query also some other rest services, and they are just fine, they have a content type. 

We need to have some Content-Type if the body is not 0, if it is 0, then it's ok to not have Content-Type.

I have been struggling with this for couple of days, and there no way i could compress payload. Even enabling compress everthing/everthing doesn't work - since there nothing in Content-Type. 

I think that this could increase greatly DP performance, bigger payloads are compressed from 6,5MB to 400kb (gzip). Brotli is even better, using Fiddler..

**Reply by rockfordlhotka** [reply]
The http channel transfers a binary (byte array) payload. What do you think is the right content-type for that data?

Also, the http channel _can_ run in "text mode" to support Blazor, so I suppose that would be a different content-type. It is the same byte array, but base64 encoded.

**Reply by BlagoCuljak** [reply]
I think that _application/octet-stream_ is better, but I can test both of them and deploy to see differences.

[https://stackoverflow.com/questions/6783921/which-mime-type-to-use-for-a-binary-file-thats-specific-to-my-program](https://stackoverflow.com/questions/6783921/which-mime-type-to-use-for-a-binary-file-thats-specific-to-my-program)

**Reply by BlagoCuljak** [reply]
I will try tomorrow, work in Europe is done. If you have any advice on code or try to add something in CSLA repo, do not hesitate to write.

**Reply by rockfordlhotka** [reply]
@BlagoCuljak I did open a work item for this (#1700), but would like to hear your results first so we have a good idea of exactly what needs to be changed.

**Reply by BlagoCuljak** [reply]
Hi Rocky, I have tried to set content type to some MIME type but with no success.

I used ProjectTracker, kickedout nuget packages, and added CSLA.dll and CSLA.Xaml.dll as reference.

Went to line you suggested, and added two headers:
![code](https://user-images.githubusercontent.com/34435929/85120947-425a4380-b224-11ea-8c80-b14588806447.PNG)

And that's from client side, and my request is good. I see those headers.
![fetch](https://user-images.githubusercontent.com/34435929/85121029-61f16c00-b224-11ea-95bb-90a6e903b10b.PNG)

But I can't set content-type from client, the server needs to respond with that info. If I set content-type from client then I get error, misused header. Of course, client should not ask for content-type, rather the server returns this info.

Have a look for eg. headers from youtube.
![youtube](https://user-images.githubusercontent.com/34435929/85121192-a54bda80-b224-11ea-83de-a0d5a768cd92.PNG)


If you have any other suggestions, feel free to contact me.

**Reply by rockfordlhotka** [reply]
Good point!

You can try setting it on the server, because there's a virtual method you can override to manipulate the reponse.

**Reply by BlagoCuljak** [reply]
What file exactly? Sorry, didn't had much experience with CSLA source.

**Reply by rockfordlhotka** [reply]
[HttpPortalController](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.Web.Mvc.Shared/Server/Hosts/HttpPortalController.cs)

On your app server you created an endpoint by creating a controller that inherits from the CSLA base type. In your controller you can override various methods to alter behavior, and I'm pretty sure there's a method where you can alter the response (like adding a header).

**Reply by BlagoCuljak** [reply]
I'm off to weekend, if you have the time, please try, and let me know the results.

**Reply by BlagoCuljak** [reply]
Off to next week!

I have tried adding headers to controller without success.  I more confused this week then the last, since all dataportal calls are POST, and when I try:
`response.Headers.Add("Content-type", "application/octet-stream");`
nothing happens. 

If you have an idea, or solution I can test before final push, please do share.

**Reply by rockfordlhotka** [reply]
When you say "nothing happens" do you mean no compression? Or that the header doesn't end up in the response?

**Reply by BlagoCuljak** [reply]
Header still the same, no Content Type on response.

**Reply by BlagoCuljak** [reply]
Hey Rocky, are you making any progress on this? I gave it one more try today, but still nothing.

What really confuses me is why CSLA is using POST all the time. And how can we get the Content-Type in response when everything is POST? Or am missing something.

If you have a code suggestion, I have the infrastructure to test it.

**Reply by rockfordlhotka** [reply]
No, I have had no time for this, and won't until _maybe_ sometime next week - my day job is really consuming right at the moment.

**Reply by BlagoCuljak** [reply]
Sure man. Let me know if I can help somehow.

**Reply by rockfordlhotka** [reply]
The data portal does always use POST - that is the correct verb for what the data portal is doing - posting a request and potentially getting a result.

**Reply by rockfordlhotka** [reply]
Does this offer any insights?

https://docs.microsoft.com/en-us/aspnet/core/performance/response-compression?view=aspnetcore-3.1

**Reply by BlagoCuljak** [reply]
I have added this to AppServerCore startup, but no difference.

Also, in that link, its being explained about MIME ContentType,and how to compress it. But all CSLA responses are not any MIME type.

**Reply by rockfordlhotka** [reply]
Have you created a subclass of the data portal controller and added elements to the header?

**Reply by BlagoCuljak** [reply]
I have just tried to add to existing class:
request.Headers.Add("context-type", "application/binary");

Didnt add any other code, thought it would work like client side...

**Reply by rockfordlhotka** [reply]
You'd need to set the response object, not the request.

**Reply by BlagoCuljak** [reply]
Hm, i was writing from my smartphone, don't recall exactly that snippet. Let me try in the morning, and get back to you.

**Reply by BlagoCuljak** [reply]
@rockfordlhotka I have added response, not the request, no diff. Here the entire [controller](https://pastebin.com/XmgktXBE), which I tried to modify.

**Reply by BlagoCuljak** [reply]
Hi Rocky, we were exploring some cloud rollout for our app, and we have like huge responses now, like 30MB of data being transferred over network. When this is compressed, its 1MB.

So, we implemented compression and it seems that its working. Headers are sent and received, so far so good.
![req res](https://user-images.githubusercontent.com/34435929/97580184-462c7c80-19f3-11eb-9d2e-b670825ed9ca.jpg)

We have tested our app in WPF, and its 5 time faster on lists.This has potential of having CSLA dramatically improve performance, like 3-5 time better, on large responses.

But Blazor gives us error on start. We made changes to HttpProxy and HttpPortalControler to make the changes work in WPF. Is there something we should else do to get the same goodies in Blazor?
![error](https://user-images.githubusercontent.com/34435929/97594364-e8079580-1a02-11eb-8d1c-a6e6a3988a59.jpg)

I don't get from where does this request comes?

**Reply by rockfordlhotka** [reply]
What specifically did you write in `HttpProxy`? Perhaps it is something not supported by WebAssembly or .NET on wasm?

There are absolutely some limitations on the HttpClient type in wasm, including the inability to transfer binary data. Which is why it is necessary to have a different server endpoint for Blazor that expects text encoding instead of binary. [For example](https://github.com/MarimerLLC/csla/blob/a2c2494b00969419335b1a1943e46324ec76de27/Samples/BlazorExample/BlazorExample/Server/Controllers/DataPortalController.cs#L12):

```c#
  [Route("api/[controller]")]
  [ApiController]
  public class DataPortalController : Csla.Server.Hosts.HttpPortalController
  {
    public DataPortalController()
    {
      UseTextSerialization = true;
    }

    [HttpGet]
    public string Get()
    {
      return "Running";
    }

    public override Task PostAsync([FromQuery] string operation)
    {
      return base.PostAsync(operation);
    }
```

**Reply by BlagoCuljak** [reply]
Hi Rocky, we have changed 3 things:
1. HttpPortalController, we added response Headers
2. HttpProxy, add Header we accept, and decompression:

https://github.com/BlagoCuljak/csla/commit/99d9ac6ed8340bb7c5587dd443e7f72e030152ee#diff-cb2884b6b674fcce4d88988f77f0772450d36535a6179856705276608d1d3da8

3. Changed Startup.cs in AppServerCore (commented although, you can uncomment it):
https://github.com/MarimerLLC/csla/commit/75ed87455517d7f69a098a73cf0e1d13268f9117

So, we tracked down DataPortalClients and they all seems to work, they use HttpProxy. We see that Blazor Server side is making request and Dataportal has all correct response. But we still have error on Blazor and Blazor app is not served at all.

We think it's due to fact that now Blazor doesn't decompress the payload. Where should we search for Blazor equivalent of HttpProxy?

**Reply by rockfordlhotka** [reply]
Blazor uses the normal HttpProxy, but it might use a different code path because `UseTextSerialization` will be `true`.

**Reply by BlagoCuljak** [reply]
We used Server side Blazor, it didn't used text serilization, and the breakpoints haven't been hit when debuging Blazor. Wpf would hit HttpProxy.

Still,the performance gain is just great news, hope to finish this off and make pull request, we need this so badly...

**Reply by rockfordlhotka** [reply]
Oh, yeah, server-side Blazor would use the normal HttpProxy.

**Reply by rockfordlhotka** [reply]
Looking at the graphic you posted, doesn't it look like the data portal request is coming from a browser?

I guess I've never looked at the traffic between a web server and app server. Does the web server make itself look like a browser when calling the app server?

**Reply by BlagoCuljak** [reply]
Hi Rocky,
The Blazor requests data as as any other UI framework supported by CSLA. In Fiddler, you can see localhost:44386 (Blazor), and blagoc:8050 (requests to dataportal server).

[Screen Recording (3.11.2020. 9-55-17).zip](https://github.com/MarimerLLC/csla/files/5480278/Screen.Recording.3.11.2020.9-55-17.zip)


Using Blazor, I get this error on login, and on every other page, I remeber, this same error was also in WPF before we implemented decompression:
```
Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware: Error: An unhandled exception has occurred while executing the request.

System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'Unhandled CSLA Known type was found')
   at Csla.DataPortal`1.Fetch(Type objectType, Object criteria)
   at Csla.DataPortal`1.Fetch(Object[] criteria)
   at Csla.DataPortal.Fetch[T](Object[] criteria)
   at FactorOne.Library.Security.CustomIdentity.GetCustomIdentity(String username, String password) in C:\Development\FactorOne\FactorOne.BusinessLibrary.Shared\Security\CustomIdentity.cs:line 97
   at FactorOne.Ui.Blazor.Areas.Identity.Pages.Account.LoginModel.OnPostAsync() in C:\Development\FactorOne\FactorOne.Ui.Blazor.Server\Areas\Account\Pages\Login.cshtml.cs:line 59
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.ExecutorFactory.GenericTaskHandlerMethod.Convert[T](Object taskAsObject)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.ExecutorFactory.GenericTaskHandlerMethod.Execute(Object receiver, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeHandlerMethodAsync()
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeNextPageFilterAsync()
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.Rethrow(PageHandlerExecutedContext context)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure.PageActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)
```

**Reply by rockfordlhotka** [reply]
I thought you were using server-side Blazor?

**Reply by BlagoCuljak** [reply]
I dont get it, those are server side Blazor requests.

**Reply by rockfordlhotka** [reply]
Yeah, so the client (browser) shouldn't matter here at all. The browser communicates with the web server via SignalR and won't interact with anything other than the web server.

So the code calling the data portal (and thus app server) is running on the web server in ASP.NET Core. You should be able to set breakpoints and step through any of that code, because it is just normal web server code.

**Reply by BlagoCuljak** [reply]
Hm, let me clean project, delete .vs and give it another try in the morning.

**Reply by BlagoCuljak** [reply]
Hi Rocky, you were right, now we can see what was wrong, main problem was that Csla in Blazor was from NuGet, not our changed Csla.dll, that's why it didn't worked first time.

Is it ok, if I make [pull request with changes seen here?](https://github.com/BlagoCuljak/csla/commit/99d9ac6ed8340bb7c5587dd443e7f72e030152ee#diff-cb2884b6b674fcce4d88988f77f0772450d36535a6179856705276608d1d3da8)

**This should not in any way** affect existing Csla installations, since you need the code in [AppServerCore Startup.cs](https://github.com/MarimerLLC/csla/commit/75ed87455517d7f69a098a73cf0e1d13268f9117) to enable compression.

If you accept pull request, when do you think new version could be available? We want to deploy it to cloud as soon as possible.

**Reply by rockfordlhotka** [reply]
@BlagoCuljak I'm good with a PR, this can go into 5.4.0.

I would like a better explanation of all the elements added to the header - specifically the stuff around image formats?

**Reply by rockfordlhotka** [reply]
Also, wouldn't you need to enable compression on the client too? You'd want compression for data flowing from client to the app server?

**Reply by rockfordlhotka** [reply]
(and I'll need a contributor agreement from you)

**Reply by rockfordlhotka** [reply]
https://github.com/MarimerLLC/csla/issues/1881

**Reply by BlagoCuljak** [reply]
Hi Rocky,
Image formats are application/octet-stream, that is for DP, as we discussed, and all others are coming from Blazor, either Server or WebAssembly. They can be observed on any Blazor app.

I will make the adjustments accordingly, although you can define headers in dataportal controller to override our defined headers, we wanted to have a broad range of header types, that can cover different scenario.

As for client side, the update operations all have compression, and it seems to work. This is also major part of improvements, working with documents that have hundreds of child lines, the performance gain is real nice! Uploading the payload in ADSL connection is way, way faster.
![update](https://user-images.githubusercontent.com/34435929/98207612-3a354300-1f3c-11eb-955c-557eede8c877.png)

The other changes are pretty self-explanatory, automatic decompression on client, and enabling compression in AppServerCore. If you think it's better, I can also push the changes in AppServerCore, but keep them commented, so we can have easy explanation for CSLA users.

I hope that you will have prerelease v5.4 so we can have this change in this calendar year. That is very important to us.

Sent signed agreement and [made pull request](https://github.com/MarimerLLC/csla/pull/1882).

**Comment by rockfordlhotka** [comment, marked as answer]
https://github.com/MarimerLLC/csla/issues/1700
